### PR TITLE
Add a hook to manage permissions

### PR DESCRIPTION
**Pull request by qsomazzi**
This is a first draft, improvements and/or advices are welcome ! :)

### CODE REVIEW

**Comment by Soullivaneuh** [comment]
I don't really like this big `canAccess` method.

What is the goal of that?

Why not using `@Security` tag or another security component stuff?

**Comment by Koc** [comment]
IMHO better use `denyAccessUnlessGranted` shortcut http://symfony.com/blog/new-in-symfony-2-6-new-shortcut-methods-for-controllers with fallback for older Sf versions

**Comment by Koc** [comment]
@Soullivaneuh `@Security` is worse because in this case we should add one more dependency: for framework extra bundle.

**Comment by Soullivaneuh** [comment]
Agree with @Koc.

**Comment by jvasseur** [comment]
@Koc We can't use the denyAccessUnlessGranted method because we need to call the Admin class isGranted method.

**Comment by Soullivaneuh** [comment]
@jvasseur Could not this admin class be hooked by security component?

**Comment by qsomazzi** [comment]
@Koc @Soullivaneuh We can't use `denyAccessUnlessGranted` because is sf2.6 or Sonata admin supports sf2.3.

The goal of this PR is only to set a small Hook to easily extends permissions management. In my project I need to check a specific permissions not manageable with `admin->isGranted` and with this `canAccess` method I can override the check easily.
This shouldn't supposed to be BC.

**Comment by jvasseur** [comment]
@Soullivaneuh This would would be strange since the Admin class is calling the security component most of the time. And this could lead to an infinite call loop.

**Comment by Soullivaneuh** [comment]
> The goal of this PR is only to set a small Hook to easily extends permissions management.
> This shouldn't supposed to be BC.

I know, but what do you mean? Why not using directly `isGranted` instead of making a big switch case?

**Comment by qsomazzi** [comment]
At first, I thought about create something like that `protected function isGranted($role, $object = null);`
But after a discussion with @rande, we agree that it's better to have only one entry point and besides some actions check the same role check.

**Comment by Soullivaneuh** [comment]
Ok @qsomazzi what I don't really like is calling this method is each action and passing... the action name.

I don't know. Not sure if I'm right and if it's possible, but this could be a place for a listener and security voter. What do you think?

**Comment by qsomazzi** [comment]
@Soullivaneuh can you come to Gitter to discuss about it ?

**Comment by Soullivaneuh** [comment]
Of course, I didn't know we have a Gitter for Sonata.

Sullivan SENECHAL
Le 23 sept. 2015 10:45, "Quentin SOMAZZI" <notifications@github.com> a
écrit :

**Comment by qsomazzi** [comment]
@Soullivaneuh I've simplified the check and make it more extendable.
I don't think a listener or voter is worth it here.

Maybe change to pass `acl` instead of `aclAction` if you think it's clearer.

File: Controller/CRUDController.php [hunk 1]
```diff
@@ -1298,6 +1276,33 @@ protected function getCsrfToken($intention)
...
 
     /**
+     * Hook to handle access authorization.
+     *
+     * @param string $action
+     * @param mixed  $object
+     */
+    protected function checkAccess($action, $object = null)
+    {
+        $access = array_merge($this->accessMapping, array(
```
Comment (rande) [inline Controller/CRUDController.php]: This should be the other way around

File: Controller/CRUDController.php [hunk 2]
```diff
@@ -1298,6 +1276,33 @@ protected function getCsrfToken($intention)
...
+            'aclAction'                     => 'MASTER',
```
Comment (rande) [inline Controller/CRUDController.php]: action prefix can be removed

Comment (Soullivaneuh) [inline Controller/CRUDController.php]: You mean suffix. :-)

**Comment by qsomazzi** [comment]
@rande done

**Comment by Bladrak** [comment]
@qsomazzi don't forget to document this

File: Controller/CRUDController.php [hunk 3]
```diff
@@ -1298,6 +1276,33 @@ protected function getCsrfToken($intention)
...
     /**
+     * Hook to handle access authorization.
+     *
+     * @param string $action
+     * @param mixed  $object
+     */
+    protected function checkAccess($action, $object = null)
+    {
+        $access = array_merge(array(
+            'acl'                     => 'MASTER',
```
Comment (rande) [inline Controller/CRUDController.php]: add support for array

**Comment by rande** [comment]
Look almost of for me. Just add documentation and the array support.

**Comment by qsomazzi** [comment]
@Bladrak @rande @Soullivaneuh Doc added

**Comment by Soullivaneuh** [comment]
Just a question @qsomazzi @rande: What happen if we want to add extra condition rather than a role label?

**Comment by qsomazzi** [comment]
if you want an extra condition, i think you will have to override the method then call the parent.
You can't add an extra condition with this for the moment

**Comment by qsomazzi** [comment]
@Soullivaneuh @rande is it ok for you ? can we merge it ?

File: Controller/CRUDController.php [hunk 4]
```diff
@@ -43,6 +43,8 @@ class CRUDController extends Controller
      */
     protected $admin;
```
Comment (Soullivaneuh) [inline Controller/CRUDController.php]: Missing php doc block.

File: Controller/CRUDController.php [hunk 5]
```diff
@@ -1298,6 +1276,51 @@ protected function getCsrfToken($intention)
     }
 
     /**
+     * Hook to handle access authorization.
+     *
+     * @param string $action
+     * @param mixed  $object
```
Comment (Soullivaneuh) [inline Controller/CRUDController.php]: `mixed` -> `object`?

File: Controller/CRUDController.php [hunk 6]
```diff
@@ -1298,6 +1276,51 @@ protected function getCsrfToken($intention)
...
+
+        if (array_key_exists($action, $access)) {
+            if (is_array($access[$action])) {
+                foreach ($access[$action] as $role) {
+                    if (false === $this->admin->isGranted($role, $object)) {
+                        $isGranted = false;
+                    }
+                }
+            } else {
+                if (false === $this->admin->isGranted($access[$action], $object)) {
```
Comment (Soullivaneuh) [inline Controller/CRUDController.php]: I think this could be simplified with an `elseif`.

File: Controller/CRUDController.php [hunk 7]
```diff
@@ -1298,6 +1276,51 @@ protected function getCsrfToken($intention)
...
+                    $isGranted = false;
+                }
+            }
+        } else {
+            $isGranted = false;
+        }
+
+        if (!$isGranted) {
```
Comment (Soullivaneuh) [inline Controller/CRUDController.php]: `false === $isGranted`

File: Controller/CRUDController.php [hunk 8]
```diff
@@ -1298,6 +1276,51 @@ protected function getCsrfToken($intention)
...
+            throw new AccessDeniedException();
```
Comment (Soullivaneuh) [inline Controller/CRUDController.php]: I think you should add an explicit exception message to help developers.

**Comment by Soullivaneuh** [comment]
Is I understand well, `checkAccess` will always return true if action is not found, isn't it?

This is not really secure IMO. If not, why not set `$isGranted` to `false` as default?

**Comment by qsomazzi** [comment]
No this will return `false` because if the action isn't in the array, `array_key_exist` will return `false`,  so `Access Denied`

**Comment by qsomazzi** [comment]
Code on diet (thanks to @Bladrak), more understandable @Soullivaneuh

File: Controller/CRUDController.php [hunk 9]
```diff
@@ -1298,6 +1281,43 @@ protected function getCsrfToken($intention)
...
+            throw new AccessDeniedException(sprintf('Access Denied to the action %s', $action));
+        }
+
+        if (!is_array($access[$action])) {
+            $access[$action] = array($access[$action]);
+        }
+
+        foreach ($access[$action] as $role) {
+            if (false === $this->admin->isGranted($role, $object)) {
+                throw new AccessDeniedException(sprintf('Access Denied to the action %s and roles [%s]', $action, implode($access[$action], ', ')));
```
Comment (rande) [inline Controller/CRUDController.php]: you should print only the current role.

**Comment by rande** [comment]
Not sure if the ``accessMapping`` property should be part of the Admin class or the controller.

**Comment by Soullivaneuh** [comment]
> Not sure if the ``accessMapping`` property should be part of the Admin class or the controller. 

I would say admin class so we wont be forced to extends a controller just for a property.

File: Controller/CRUDController.php [hunk 10]
```diff
@@ -1298,6 +1281,43 @@ protected function getCsrfToken($intention)
...
+            'history'                 => 'EDIT',
+            'edit'                    => 'EDIT',
+            'show'                    => 'VIEW',
+            'create'                  => 'CREATE',
+            'delete'                  => 'DELETE',
+            'batchDelete'             => 'DELETE',
+            'list'                    => 'LIST',
+        ), $this->accessMapping);
+
+        if (!array_key_exists($action, $access)) {
```
Comment (Koc) [inline Controller/CRUDController.php]: It looks more like `InvalidArgumentException` here or `LogicException` - that something is missconfigured.

**Comment by rande** [comment]
@Soullivaneuh yes

**Comment by qsomazzi** [comment]
It's done

**Comment by rande** [comment]
@qsomazzi it also means the checkAccess can be move to the admin class.

File: Controller/CRUDController.php [hunk 11]
```diff
@@ -1298,6 +1274,43 @@ protected function getCsrfToken($intention)
...
+            'edit'                    => 'EDIT',
+            'show'                    => 'VIEW',
+            'create'                  => 'CREATE',
+            'delete'                  => 'DELETE',
+            'batchDelete'             => 'DELETE',
+            'list'                    => 'LIST',
+        ), $this->admin->getAccessMapping());
+
+        if (!array_key_exists($action, $access)) {
+            throw new AccessDeniedException(sprintf('Access Denied to the action %s', $action));
```
Comment (eko) [inline Controller/CRUDController.php]: I think this is more an `InvalidArgumentException` than an `AccessDeniedException` and should display a message which will help developers. Something like `Action "%s" could not be found in access mapping. Please make sure your action is defined into your admin class getAccessMapping() method.`

**Comment by Soullivaneuh** [comment]
> @qsomazzi it also means the checkAccess can be move to the admin class.

Not sure about that. Can you elaborate?

**Comment by rande** [comment]
@Soullivaneuh route definitions are also part of the Admin, so it also make sense to have one place to check access to an action

**Comment by qsomazzi** [comment]
@rande done

**Pull request merged**